File: app/src/main/java/com/prey/activities/BarcodeActivity.java
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2016 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.activities;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.View;
import android.widget.CompoundButton;
import android.widget.TextView;

import com.google.android.gms.common.api.CommonStatusCodes;
import com.google.android.gms.vision.barcode.Barcode;
import com.prey.PreyAccountData;
import com.prey.PreyConfig;
import com.prey.PreyLogger;
import com.prey.PreyUtils;
import com.prey.R;
import com.prey.barcode.BarcodeCaptureActivity;
import com.prey.net.PreyWebServices;

public class BarcodeActivity extends Activity implements View.OnClickListener {


    private CompoundButton autoFocus;
    private CompoundButton useFlash;
    private TextView statusMessage;
    private TextView barcodeValue;

    private static final int RC_BARCODE_CAPTURE = 9001;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_barcode);

        statusMessage = (TextView) findViewById(R.id.status_message);
        barcodeValue = (TextView) findViewById(R.id.barcode_value);

        autoFocus = (CompoundButton) findViewById(R.id.auto_focus);
        useFlash = (CompoundButton) findViewById(R.id.use_flash);

        findViewById(R.id.read_barcode).setOnClickListener(this);
    }


    @Override
    public void onClick(View v) {
        if (v.getId() == R.id.read_barcode) {

            Intent intent = new Intent(this, BarcodeCaptureActivity.class);
            intent.putExtra(BarcodeCaptureActivity.AutoFocus, autoFocus.isChecked());
            intent.putExtra(BarcodeCaptureActivity.UseFlash, useFlash.isChecked());

            startActivityForResult(intent, RC_BARCODE_CAPTURE);
        }

    }


    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == RC_BARCODE_CAPTURE) {
            if (resultCode == CommonStatusCodes.SUCCESS) {
                if (data != null) {
                    Barcode barcode = data.getParcelableExtra(BarcodeCaptureActivity.BarcodeObject);
                    statusMessage.setText(R.string.barcode_success);
                    barcodeValue.setText(barcode.displayValue);
                    String barcodeValue = barcode.displayValue;
                    PreyLogger.d("Barcode read: " + barcodeValue);

                    if (barcodeValue.indexOf("prey") >= 0) {

                        barcodeValue = barcodeValue.substring(5);
                        String[] pairs = barcodeValue.split("&");
                        String apikey = "";
                        String mail = "batch@preyproject.com";
                        for (String pair : pairs) {
                            String[] llave = pair.split("=");
                            PreyLogger.i("key[" + llave[0] + "]" + llave[1]);
                            if (llave[0].equals("api_key")) {
                                apikey = llave[1];
                            }

                        }
                        if (!"".equals(apikey)) {
                            new AddDeviceToApiKeyBatch().execute(apikey, mail, PreyUtils.getDeviceType(this));
                        }


                    }

                } else {
                    statusMessage.setText(R.string.barcode_failure);
                    PreyLogger.d("No barcode captured, intent data is null");
                }
            } else {
                statusMessage.setText(String.format(getString(R.string.barcode_error),
                        CommonStatusCodes.getStatusCodeString(resultCode)));
            }
        } else {
            super.onActivityResult(requestCode, resultCode, data);
        }
    }

    String error = null;

    private static final int NO_MORE_DEVICES_WARNING = 0;
    private static final int ERROR = 3;

    private boolean noMoreDeviceError = false;

    private class AddDeviceToApiKeyBatch extends AsyncTask<String, Void, Void> {
        @Override
        protected void onPreExecute() {

        }

        @Override
        protected Void doInBackground(String... data) {
            try {
                error = null;
                Context ctx = getApplicationContext();

                PreyLogger.i("apikey:" + data[0] + " mail:" + data[1] + " device:" + data[2]);

                // if(!PreyConfig.getPreyConfig(ctx).isThisDeviceAlreadyRegisteredWithPrey()) {
                PreyAccountData accountData = PreyWebServices.getInstance().registerNewDeviceWithApiKeyEmail(ctx, data[0], data[1], data[2]);
                if (accountData != null) {
                    PreyConfig.getPreyConfig(ctx).saveAccount(accountData);
                    PreyConfig.getPreyConfig(getApplicationContext()).saveAccount(accountData);
                    PreyConfig.getPreyConfig(getApplication()).registerC2dm();
                    PreyWebServices.getInstance().sendEvent(getApplication(), PreyConfig.ANDROID_SIGN_IN);
                }

            } catch (Exception e) {
                error = e.getMessage();
            }
            return null;
        }

        @Override
        protected void onPostExecute(Void unused) {
            PreyLogger.i("error[" + error + "]");
            if (error == null) {
                String message = getString(R.string.device_added_congratulations_text);
                Bundle bundle = new Bundle();
                bundle.putString("message", message);
                PreyConfig.getPreyConfig(getApplicationContext()).setCamouflageSet(true);
                Intent intent = new Intent(getApplicationContext(), PermissionInformationActivity.class);
                intent.putExtras(bundle);
                startActivity(intent);

                finish();


            } else {
                showDialog(ERROR);
            }
        }
    }


    @Override
    protected Dialog onCreateDialog(int id) {
        Dialog pass = null;
        switch (id) {

            case ERROR:
                return new AlertDialog.Builder(BarcodeActivity.this).setIcon(R.drawable.error).setTitle(R.string.error_title).setMessage(error)
                        .setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int which) {
                            }
                        }).setCancelable(false).create();

            case NO_MORE_DEVICES_WARNING:
                return new AlertDialog.Builder(BarcodeActivity.this).setIcon(R.drawable.info).setTitle(R.string.set_old_user_no_more_devices_title).setMessage(error)
                        .setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {

                            public void onClick(DialogInterface dialog, int which) {
                            }
                        }).setCancelable(false).create();
        }
        return pass;
    }

    @Override
    protected void onPrepareDialog(int id, Dialog dialog) {
        AlertDialog ad = null;
        switch (id) {

            case ERROR:
                ad = (AlertDialog) dialog;
                ad.setIcon(R.drawable.error);
                ad.setTitle(R.string.error_title);
                ad.setMessage(error);
                ad.setButton(DialogInterface.BUTTON_POSITIVE, this.getString(R.string.ok), new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        // Handler code
                    }
                });

                ad.setCancelable(false);

                break;

            case NO_MORE_DEVICES_WARNING:
                ad = (AlertDialog) dialog;
                ad.setIcon(R.drawable.info);
                ad.setTitle(R.string.set_old_user_no_more_devices_title);
                ad.setMessage(error);
                ad.setButton(DialogInterface.BUTTON_POSITIVE, this.getString(R.string.ok), new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        // Handler code
                    }
                });
                ad.setCancelable(false);

                break;
            default:
                super.onPrepareDialog(id, dialog);
        }
    }
}

File: app/src/main/java/com/prey/activities/BarcodeActivity.kt
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2025 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.activities

import android.Manifest
import android.app.ProgressDialog
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.camera.mlkit.vision.MlKitAnalyzer
import androidx.camera.view.CameraController.COORDINATE_SYSTEM_VIEW_REFERENCED
import androidx.camera.view.LifecycleCameraController
import androidx.camera.view.PreviewView
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat

import com.google.mlkit.vision.barcode.BarcodeScanner
import com.google.mlkit.vision.barcode.BarcodeScannerOptions
import com.google.mlkit.vision.barcode.BarcodeScanning
import com.google.mlkit.vision.barcode.common.Barcode
import com.prey.PreyConfig
import com.prey.PreyLogger
import com.prey.R
import com.prey.barcodereader.QrCodeViewModel

import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

class BarcodeActivity : AppCompatActivity() {

    private lateinit var cameraExecutor: ExecutorService
    private lateinit var barcodeScanner: BarcodeScanner

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_barcode)
        if (!PreyConfig.getInstance(applicationContext).isTest()) {
            // Request camera permissions
            if (allPermissionsGranted()) {
                startCamera()
            } else {
                ActivityCompat.requestPermissions(
                    this, REQUIRED_PERMISSIONS, REQUEST_CODE_PERMISSIONS
                )
            }
            cameraExecutor = Executors.newSingleThreadExecutor()
        }
        PreyConfig.getInstance(applicationContext).setActivityView(ACTIVITY_BARCODE)
    }

    private fun startCamera() {
        val cameraController = LifecycleCameraController(baseContext)
        val previewView: PreviewView = findViewById(R.id.viewFinder)
        val options = BarcodeScannerOptions.Builder()
            .setBarcodeFormats(Barcode.FORMAT_QR_CODE)
            .build()
        barcodeScanner = BarcodeScanning.getClient(options)
        cameraController.setImageAnalysisAnalyzer(
            ContextCompat.getMainExecutor(this),
            MlKitAnalyzer(
                listOf(barcodeScanner),
                COORDINATE_SYSTEM_VIEW_REFERENCED,
                ContextCompat.getMainExecutor(this)
            ) { result: MlKitAnalyzer.Result? ->
                val barcodeResults = result?.getValue(barcodeScanner)
                if ((barcodeResults == null) ||
                    (barcodeResults.size == 0) ||
                    (barcodeResults.first() == null)
                ) {
                    previewView.overlay.clear()
                    previewView.setOnTouchListener { _, _ -> false }
                    return@MlKitAnalyzer
                }
                val qrCodeViewModel = QrCodeViewModel(barcodeResults[0])
                PreyLogger.d("qrCodeViewModel qrContent: ${qrCodeViewModel.qrContent}")
                registerNewDevice(qrCodeViewModel.qrContent)
                previewView.overlay.clear()
            }
        )
        cameraController.bindToLifecycle(this)
        previewView.controller = cameraController
    }

    private fun registerNewDevice(qrContent: String) {
        if (qrContent.contains("prey?api_key=")) {
            val apikey = qrContent.substring(13)
            PreyLogger.i("___apikey:$apikey")
            val progressDialog = ProgressDialog(this@BarcodeActivity)
            progressDialog.setMessage(getText(R.string.set_old_user_loading).toString())
            progressDialog.isIndeterminate = true
            progressDialog.setCancelable(false)
            progressDialog.show()
            var error: String? = null
            try {
                PreyConfig.getInstance(applicationContext).registerNewDeviceWithApiKey(apikey)
            } catch (e: Exception) {
                PreyLogger.e("Error:${e.message}", e)
                error = e.message
            }
            try {
                progressDialog.dismiss()
            } catch (e: Exception) {
                PreyLogger.e("Error:${e.message}", e)
            }
            if (error == null) {
                val message = getString(R.string.device_added_congratulations_text)
                val bundle = Bundle()
                bundle.putString("message", message)
                bundle.putString(CheckPasswordHtmlActivity.NEXT_URL, "/activation")
                PreyConfig.getInstance(applicationContext).setCamouflageSet(true)
                var intent: Intent? = null
                intent = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    Intent(applicationContext, JumpActivity::class.java)
                } else {
                    Intent(applicationContext, PermissionInformationActivity::class.java)
                }
                intent.putExtras(bundle)
                startActivity(intent)
                finish()
            } else {
                showDialog(ERROR)
            }
        }

    }

    private fun allPermissionsGranted() = REQUIRED_PERMISSIONS.all {
        ContextCompat.checkSelfPermission(
            baseContext, it
        ) == PackageManager.PERMISSION_GRANTED
    }

    override fun onDestroy() {
        super.onDestroy()
        if (!PreyConfig.getInstance(applicationContext).isTest()) {
            cameraExecutor.shutdown()
            barcodeScanner.close()
        }
    }

    companion object {
        const val ACTIVITY_BARCODE: String = "ACTIVITY_BARCODE"
        private const val ERROR = 3
        private const val REQUEST_CODE_PERMISSIONS = 10
        private val REQUIRED_PERMISSIONS =
            mutableListOf(
                Manifest.permission.CAMERA
            ).toTypedArray()
    }

    override fun onRequestPermissionsResult(
        requestCode: Int, permissions: Array<String>, grantResults:
        IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == REQUEST_CODE_PERMISSIONS) {
            if (allPermissionsGranted()) {
                startCamera()
            } else {
                Toast.makeText(
                    this,
                    R.string.permissions_not_granted_by_user,
                    Toast.LENGTH_SHORT
                ).show()
                finish()
            }
        }
    }

}